src/kalshi.rs

```diff
@@ -0,0 +1,377 @@
+    }
+
+    pub fn get_market(&self) -> &Market {
+        self.event
+            .markets
+            .iter()
+            .find(|market| market.ticker_name == self.id)
+            .with_context(|| format!("Could not find market in series {} with ticker_name {}", self.id, self.id))
+            .unwrap()
+    }
+
+    pub fn is_resolved(&self) -> bool {
+        self.get_market().status == Status::Finalized
+    }
+
+    pub fn is_active(&self) -> bool {
+        self.get_market().status == Status::Active
+    }
+
+    pub fn is_series(&self) -> bool {
+        self.event.markets.len() > 1
+    }
+
+    pub fn time_to_resolution(&self) -> Duration {
+        self.get_market().expiration_date - Utc::now()
+    }
+
+    pub fn full_url(&self) -> String {
+        // TODO: grab base from config (consistent with manifold)?
+        format!("https://kalshi.com/markets/{}", self.event.series_ticker)
+    }
+
+    pub fn title(&self) -> String {
+        self.get_market().title.clone()
+    }
+
+    pub fn get_criteria_and_sources(&self) -> String {
+        format!("{}{}", self.format_underlying_rulebook_variables(), self.get_resolution_sources_markdown())
+    }
+
+    pub fn format_underlying_rulebook_variables(&self) -> String {
+        let mut return_string = self.event.underlying.clone();
+        // For each variable in market.rulebook_variables, substitute ||variable|| with the value of the variable
+        let rulebook = self.get_market().rulebook_variables.clone();
+        // For each key in rulebook
+        for key in rulebook.as_object().unwrap().keys() {
+            // Remove the surrounding "" marks
+            let mut replacement_value = rulebook[key].to_string();
+            replacement_value = replacement_value[1..replacement_value.len()-1].to_string();
+
+            // We've seen instances with and without spaces
+            return_string = return_string.replace(&format!("||{}||", key), &replacement_value);
+            return_string = return_string.replace(&format!("|| {} ||", key), &replacement_value);
+        }
+        return return_string
+
+    }
+
+    pub fn get_resolution_sources_markdown(&self) -> String {
+        let sources = self
+            .event
+            .settlement_sources
+            .iter()
+            .map(|source| format!("<{}>", source.url.clone()))
+            .collect::<Vec<String>>();
+
+        // Return "" if there are none
+        if sources.is_empty() {
+            return "".to_string();
+        }
+        format!("\n\n\n**Resolution sources**\n\n{}", sources.join(", "))
+    }
+
+    pub fn get_binary_resolution(&self) -> Result<Option<BinaryResolution>> {
+        if self.is_resolved() {
+            match self.get_market().result {
+                Some(KalshiResult::Yes) => Ok(Some(BinaryResolution::Yes)),
+                Some(KalshiResult::No) => Ok(Some(BinaryResolution::No)),
+                Some(KalshiResult::StillOpen) => Ok(None),
+                None => Ok(None),
```

Comment:
Is this `None` case actually possible? If not, it would probably be appropriate to panic here.